src/FromFile.jl

```diff
@@ -46,16 +86,40 @@ function from_m(m::Module, s::LineNumberNode, path::String, root_ex::Expr)
         file_module_sym = Symbol(relpath(path, pathof(root)))
     end
 
+    file_module = lazy_load_file(path, root, file_module_sym)
+    return load_symbols_from_file(file_module, file_module_sym, parse_import_stmts(root_ex))
+end
+
+function parse_import_stmts(root_ex::Expr)
+    import_exs = if root_ex.head === :block
+        filter(ex -> !(ex isa LineNumberNode), root_ex.args)
+    else
+        [root_ex]
+    end
+
+    all(ex -> ex.head === :using || ex.head === :import, import_exs) || error("expected using/import statement")
+    return import_exs
+end
+
+function lazy_load_file(path::String, root::Module, file_module_sym::Symbol)
+    return lazy_load_file(()->path, root, file_module_sym)
+end
+
+function lazy_load_file(f, root::Module, file_module_sym::Symbol)
     if isdefined(root, file_module_sym)
         file_module = getfield(root, file_module_sym)
     else
+        path = f()
         file_module = Base.eval(root, :(module $(file_module_sym); include($path); end))
-
         # In interactive sessions, track generated module using Revise.jl if Revise has been loaded
         track_modules() && track(file_module, path)
     end
+    return file_module
+end
 
-    return Expr(:block, map(import_exs) do ex
+function load_symbols_from_file(file_module::Module, file_module_sym::Symbol, @nospecialize(import_exs))
```

Comment:
Just for my own curiosity: what was the rationale for `@nospecialize` here?

Reply:
I wanted to put `Any` before `import_exs` to prevent the compiler specialize in it, but I think this is probably simpler. the `@nospecialize` macro is useful as a hint to ask the compiler not to do any specialization on the type and just interpret it.